### PR TITLE
read metrics remote write config from operator

### PR DESCRIPTION
**Pull request by ydayagi**
https://issues.redhat.com/browse/ECOPROJECT-555
read configuration from operator

Signed-off-by: Yaron Dayagi <ydayagi@redhat.com>

### CODE REVIEW

**Comment by ydayagi** [comment]
go.mod will be updated once the operator PR is merged (https://github.com/project-flotta/flotta-operator/pull/148)

File: internal/metrics/remote_write_test.go [hunk 1]
```diff
@@ -11,237 +12,324 @@ import (
...
+			// when
+			err := remoteWrite.Init(configMessage)
+
+			// then
+			Expect(err).To(BeNil())
+			Expect(remoteWrite.IsEnabled()).To(BeTrue())
+			Expect(remoteWrite.IsRunning()).To(BeTrue())
+		})
+
+		It("Update", func() {
```
Comment (jakub-dzon) [inline internal/metrics/remote_write_test.go:84]: Please add tests for disabling and enabling the feature with `Update`.

**Review by jakub-dzon — CHANGES_REQUESTED** [review]
:+1: Please just add a couple of test cases.

**Review by jakub-dzon — APPROVED** [review]

File: go.mod [hunk 2]
```diff
@@ -34,6 +34,7 @@ replace (
 	github.com/containernetworking/cni => github.com/containernetworking/cni v0.8.1
 	github.com/containers/buildah => github.com/containers/buildah v1.23.1
 	github.com/opencontainers/image-spec => github.com/opencontainers/image-spec v1.0.2-0.20210819154149-5ad6f50d6283
+	github.com/project-flotta/flotta-operator => github.com/ydayagi/k4e-operator v0.0.0-20220329090830-d066b756c24a
```
Comment (gciavarrini) [inline go.mod:37]: what's missing in the operator main branch?

**Review by machacekondra — APPROVED** [review]

File: internal/metrics/remote_write.go [hunk 3]
```diff
@@ -113,38 +108,53 @@ func (w *writeClient) Write(ctx context.Context, data []byte) error {
...
+	// use new config
+	featureDisabled := newConfig == nil || newConfig.URL == ""
+
+	if featureDisabled {
+		r.client = nil
+	} else {
+		serverURL, err := url.Parse(newConfig.URL)
 		if err != nil {
-			log.Error("failed creating metrics remote write client", err)
+			log.Error("metrics remote write configuration is invalid. Can not parse URL", err)
```
Comment (masayag) [inline internal/metrics/remote_write.go:144]: is there a need to include URL in the error message so the errored value will be known?

File: go.mod (hunk 2 above)
Comment (ydayagi) [reply go.mod:37]: my other PR :-)

File: internal/metrics/remote_write.go (hunk 3 above)
Comment (ydayagi) [reply internal/metrics/remote_write.go:144]: I rely on the fact that config manager already prints entire message

File: internal/metrics/remote_write.go [hunk 4]
```diff
@@ -153,40 +163,63 @@ func (r *RemoteWrite) Update(config models.DeviceConfigurationMessage) error {
...
+		}
+	}
+
 	for {
-		r.Write()
-		time.Sleep(r.waitInterval)
+		getConfig()
+
+		if shouldExit {
+			log.Info("metrics remote write stopped")
```
[the remark this answers is not in the document]
Comment (ydayagi) [reply internal/metrics/remote_write.go:212]: OMG. I had a return here

File: internal/metrics/remote_write.go (hunk 3 above)
Comment (ydayagi) [reply internal/metrics/remote_write.go]: added just in case...

**Review by masayag — APPROVED** [review]

**Pull request merged**